pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -99,20 +113,37 @@ func (p *Provider) Init() error {
 
 // Provide allows the consul catalog provider to provide configurations to traefik using the given configuration channel.
 func (p *Provider) Provide(configurationChan chan<- dynamic.Message, pool *safe.Pool) error {
+	var err error
+	p.client, err = createClient(p.Endpoint)
+	if err != nil {
+		return fmt.Errorf("unable to create consul client: %w", err)
+	}
+
+	if p.ConnectAware {
+		pool.GoCtx(p.watchConnectTLS)
+	}
+
 	pool.GoCtx(func(routineCtx context.Context) {
 		ctxLog := log.With(routineCtx, log.Str(log.ProviderName, "consulcatalog"))
 		logger := log.FromContext(ctxLog)
 
 		operation := func() error {
-			var err error
-
-			p.client, err = createClient(p.Endpoint)
-			if err != nil {
-				return fmt.Errorf("unable to create consul client: %w", err)
+			var (
+				err      error
+				certInfo *connectCert
+			)
+
+			// If we are running in connect aware mode then we need to
+			// make sure that we obtain the certificates before starting
+			// the service watcher, otherwise a connect enabled service
+			// that gets resolved before the certificates are available
+			// will cause an error condition.
+			if p.ConnectAware {
+				certInfo = <-p.certChan
```

Comment:
I'm a bit worried that we could be blocking forever here, if someone is careless enough to configure Traefik in connect aware mode, while the consul cluster is actually not using connect. So I think we should add a logging message right before there.
```logger.Infof("Waiting for Connect certificate before building first configuration")```